File: build-tools/codeassist-builder-plugin/src/main/java/com/tyron/builder/model/level2/GraphItem.java
package com.tyron.builder.model.level2;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * A node in a dependency graph, representing a direct or transitive dependency.
 *
 * This does not directly contain artifact information, instead it focuses on the graph
 * information (transitive dependencies) as well as the usage of this particular dependency
 * in this node of the graph (ie what are its modifiers: what version was originally requested.)
 *
 * @since 2.3
 */
public interface GraphItem {

    /**
     * Returns the artifact address in a unique way.
     *
     * This is either a module path for sub-modules, or a maven coordinate for external
     * dependencies.
     *
     * The maven coordinates are in the format: groupId:artifactId:version[:classifier][@extension]
     *
     */
    @NotNull
    String getArtifactAddress();

    /**
     * Returns this library's Maven coordinates, as requested in the build file.
     */
    @Nullable
    String getRequestedCoordinates();

    /**
     * Return the direct dependency of this node.
     */
    @NotNull
    List<GraphItem> getDependencies();
}

File: build-tools/codeassist-builder-plugin/src/main/java/com/tyron/builder/model/v2/ide/GraphItem.kt
package com.tyron.builder.model.v2.ide

import com.tyron.builder.model.v2.AndroidModel

/**
 * A node in a dependency graph, representing a direct or transitive dependency.
 *
 * This does not directly contain artifact information, instead it focuses on the graph
 * information (transitive dependencies) as well as the usage of this particular dependency
 * in this node of the graph (ie what are its modifiers: what version was originally requested.)
 *
 */
interface GraphItem: AndroidModel {
    /**
     * A Unique key representing the library, and allowing to match it with a [Library] instance
     */
    val key: String

    /**
     * Returns this library's Maven coordinates, as requested in the build file.
     */
    val requestedCoordinates: String?

    /**
     * Return the direct dependency of this node.
     */
    val dependencies: List<GraphItem>
}
